internal/jobobject/queue_test.go

```diff
@@ -0,0 +1,89 @@
+package jobobject
+
+import (
+	"testing"
+	"time"
+)
+
+func TestReadOrWait(t *testing.T) {
+	q := newMessageQueue()
+
+	// What we care about for job objects. Verify that we can just read when theres a value
+	// available in the queue, otherwise block if there's not a value and wait for one to become
+	// available. This tests that a client could start reading notifications late into the lifecycle
+	// of a job when there's likely a couple of messages already queued --> flush the queue of
+	// messages, and then wait for more to arrive.
+	go func() {
+		q.write(1)
+		q.write(2)
+		q.write(3)
+		time.Sleep(time.Second * 5)
+		q.write(4)
+	}()
+
+	time.Sleep(time.Second * 2)
+
+	for {
+		if msg, err := q.readOrWait(); err != nil {
+			t.Fatal(err)
+		} else {
+			// We successfully received the last message after the timeout.
+			if msg == 4 {
+				break
+			}
+		}
+	}
+}
+
+func TestReadOrWaitClose(t *testing.T) {
+	q := newMessageQueue()
+
+	go func() {
+		q.write(1)
+		q.write(2)
+		q.write(3)
+		time.Sleep(time.Second * 5)
+		q.close()
+	}()
+
+	time.Sleep(time.Second * 2)
+
+	for {
+		if _, err := q.readOrWait(); err != nil {
+			if err == errQueueClosed {
+				break
+			}
+			t.Fatal(err)
+		}
+	}
+}
+
+func TestReadWrite(t *testing.T) {
+	q := newMessageQueue()
+
+	// Reading from an empty queue should return errQueueEmpty
+	if _, err := q.read(); err != errQueueEmpty {
+		t.Fatal(err)
+	}
+
+	// Write 1 to the queue and read this later.
+	if err := q.write(1); err != nil {
+		t.Fatal(err)
+	}
+
+	// Read the value. Value will be dequeued.
+	if msg, err := q.read(); err != nil || msg != 1 {
+		t.Fatal(err)
+	}
+
+	// We just read a value, now try and read again and verify that we get errQueueEmpty again.
+	if _, err := q.read(); err != errQueueEmpty {
+		t.Fatal(err)
+	}
+
+	// Close the queue and verify that we get an error on write.
+	q.close()
+	if err := q.write(1); err != errQueueClosed {
+		t.Fatal(err)
+	}
+}
```

Comment:
I think it would be good to add a test case where we read all the contents in the queue, issue a new write, and read again. This will check to make sure that the ready message queue functions properly both for the first intial write and for all subsequent writes where the write queue had previously been emptied.

Reply:
That's almost what TestReadOrWait does pretty much although it doesn't block on the first read as there will be a value, but it does block waiting for the last write. I've added one more test case that will start reading before there's anything in the queue to test the ready channel for the first write and that will also drain the queue and get sent one last message after it's drained to test that case also.